frappe/erpnext
# fix: misc minor Product Bundle fixes

## Problems Addressed:

1. On creating a Purchase Order from a Sales Order that has multiple Product Bundles in it, all the Product Bundles get added to the Purchase Order, even if they weren't selected.

<details>
<summary>Testing Info</summary>

1. Create a Sales Order
2. Add one or more Product Bundles to it
3. Create > Purchase Order
4. While selecting Items in the pop-up, don't select one or more of the Product Bundles

![Screenshot 2021-11-24 at 6 36 49 AM](https://user-images.githubusercontent.com/25903035/143152892-3e6c5be6-8515-4da0-bb05-bb0dca257523.png)

_Expected Result:_ Only the child Items of the Product Bundle that was selected should show up in the Purchase Order.

![Screenshot 2021-11-24 at 6 37 35 AM](https://user-images.githubusercontent.com/25903035/143152950-0474d6be-bc2c-47fb-be44-c9db6c5cdfea.png)

</details>

2. When _Calculate Product Bundle Price based on Child Items' Rates_(under Selling Settings) is enabled, the price of Product Bundles in Quotations, Sales Orders, Delivery Notes etc should be calculated based on the rates entered in the Packed/Bundle Items table for its child Items. However, when there are multiple Product Bundles in the same Items list, this calculation is carried out incorrectly for every Product Bundle except the first.

<details>
<summary>Testing Info</summary>

1. Open Selling Settings
2. Enable _Calculate Product Bundle Price based on Child Items' Rates_
3. Create two or more Product Bundles
4. Create a Sales Order
5. Add two or more Product Bundles to the Items table
6. Save
7. Edit the rates for the child Items in the Packed Items table
8. Save

_Expected Result:_ The prices of all the Product Bundles should be equal to the sum of the prices of its child Items.

![Screenshot 2021-11-24 at 6 35 13 AM](https://user-images.githubusercontent.com/25903035/143152764-32c745a8-c856-4769-9ad5-5f78fe6c696a.png)

</details>

3. On deleting Product Bundles from the Items table of any DocType that has a Packed/Bundle Items table(Sales Orders, Invoices, etc), the indices are not reset.

<details>
<summary>Images</summary>

_Before:_

![Screenshot 2021-11-24 at 9 09 54 PM](https://user-images.githubusercontent.com/25903035/143269333-869918b4-40d8-49cd-b430-678f56f68e3c.png)

_Deleting the second row:_

![Screenshot 2021-11-24 at 9 11 53 PM](https://user-images.githubusercontent.com/25903035/143269461-27abca8b-e2a8-415a-9aab-30c41c923e1e.png)


_After:_

![Screenshot 2021-11-24 at 9 16 24 PM](https://user-images.githubusercontent.com/25903035/143270283-dd052709-5ba8-4a7c-8b3b-29b45220d05f.png)

</details>

<details>
<summary>Testing Info</summary>

1. Create 3 or more Product Bundles
2. Create a Sales Order
3. Add 3 or more Product Bundles to the Items table
4. Save
5. Delete any of the Product Bundles in the Items table(except the last)
6. Save
7. Refresh

_Expected Result:_ The indices should be displayed in order.

</details>

4. If there are more than one Product Bundles in the Items table and any of them other than the last one are deleted, this happens till you refresh the page:

![Screenshot 2021-11-29 at 6 52 48 PM](https://user-images.githubusercontent.com/25903035/144119476-4615a72c-0b8d-4184-b52f-c7e997334cab.png)

<details>
<summary>Testing Info</summary>

1. Create 3 or more Product Bundles
2. Create a Sales Order
3. Add 3 or more Product Bundles to the Items table
4. Save
5. Delete any of the Product Bundles in the Items table(except the last)
6. Save

</details>
